File: 2023/02. Cube Conundrum/c/two/src/main.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "utils.h"

#define DELIMS " :;,"
#define NUM_COLORS 3

int main(const int argc, const char** const argv)
{
    if (argc != 2)
    {
        fprintf(stderr, "Usage: %s <INPUT_FILE>\n", argv[0]);
        return 1;
    }

    FILE* const infile = fopen(argv[1], "r");
    if (infile == NULL)
    {
        perror("Failed to open input file");
        return 1;
    }

    // Puzzle result
    int powerSum = 0;

    // Read each line

    LineBuffer line;
    lineBufferCreate(&line);

    for (int lineNum = 1; readNextLine(infile, &line); ++lineNum)
    {
        // Minimum number of cubes of each color needed for current game to be possilbe
        int minNeeded[NUM_COLORS] = {0};

        // Skip first two tokens
        strtok(line.line, DELIMS);
        strtok(NULL, DELIMS);

        // Read ammount, color pairs
        for (char* ammountString = strtok(NULL, DELIMS); ammountString != NULL; ammountString = strtok(NULL, DELIMS))
        {
            const int ammount = atoi(ammountString);
            if (ammount == 0)
            {
                fprintf(stderr, "Invalid ammount \"%s\" on line %d\n", ammountString, lineNum);
                goto cleanup;
            }

            const char* const colorString = strtok(NULL, DELIMS);
            if (colorString == NULL)
            {
                fprintf(stderr, "Expected color on line %d\n", lineNum);
                goto cleanup;
            }

            // If ammount is greater than current max for given color record new max
            if (strcmp(colorString, "red") == 0)
            {
                minNeeded[0] = ammount > minNeeded[0] ? ammount : minNeeded[0];
            }
            else if (strcmp(colorString, "green") == 0)
            {
                minNeeded[1] = ammount > minNeeded[1] ? ammount : minNeeded[1];
            }
            else if (strcmp(colorString, "blue") == 0)
            {
                minNeeded[2] = ammount > minNeeded[2] ? ammount : minNeeded[2];
            }
            else
            {
                fprintf(stderr, "Unknown color \"%s\" on line %d\n", colorString, lineNum);
                goto cleanup;
            }
        }

        powerSum += minNeeded[0] * minNeeded[1] * minNeeded[2];
    }

    printf("Result: %d\n", powerSum);

cleanup:
    lineBufferDestroy(&line);

    if (fclose(infile) == EOF)
        fputs("fclose() failed\n", stderr);

    return 0;
}

File: 2023/02. Cube Conundrum/rust/two/src/main.rs
use std::env;
use std::fs;

const NUM_COLORS: usize = 3;
const COLOR_STRINGS: [&str; NUM_COLORS] = ["red", "green", "blue"];

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <INPUT_FILE>", args[0]);
        return;
    }

    // Puzzle result
    let mut power_sum = 0;

    // Read line by line
    let contents = fs::read_to_string(&args[1]).expect("Could not read input file");
    for (line_num, line) in contents.lines().enumerate() {
        // Min number of cubes of each color needed for game to be possible
        let mut min_needed = [0; NUM_COLORS];

        // Sperate line into tokens, skip the "Game" and game id tokens, and remove emtpy spaces
        let mut tokens = line
            .split(|c| " :;,".contains(c))
            .skip(2)
            .filter(|part| !part.is_empty());

        // Iterate over each ammount, color pair
        while let Some(ammount) = tokens.next() {
            // Parse the ammount
            let ammount: i32 = ammount
                .parse()
                .expect(&format!("Could not convert \"{ammount}\" to integer"));

            // Parse the color
            let color = tokens
                .next()
                .expect(&format!("Expected color on line {line_num}"));

            // Get the 0-2 index for the color
            let color = COLOR_STRINGS
                .iter()
                .enumerate()
                .find(|(_, color_string)| **color_string == color)
                .map(|(index, _)| index)
                .expect(&format!("Unknown color \"{color}\" on line {line_num}"));

                // Update min needed of ammount is greater
                min_needed[color] = min_needed[color].max(ammount);
        }

        power_sum += min_needed[0] * min_needed[1] * min_needed[2];
    }

    println!("Result: {power_sum}");
}
